pytorch_lightning/core/lightning.py

```diff
@@ -1757,3 +1700,77 @@ def get_tqdm_dict(self) -> Dict[str, Union[int, str]]:
         rank_zero_warn("`get_tqdm_dict` was renamed to `get_progress_bar_dict` in v0.7.3"
                        " and this method will be removed in v1.0.0", DeprecationWarning)
         return self.get_progress_bar_dict()
+
+    def _auto_register_hparams(self):
+        """
+        Removes the need to pass in hparams. Instead, we register every argument in init
+        to the module with some caveats:
+        1. we don't overwrite the property if it already exists
+        2. we also store a module_arguments property for model loading and saving
+        """
+        frame = inspect.currentframe()
+
+        frame_args = _collect_init_args(frame, {})
+        init_args = {k: v for k, v in frame_args.items()
+                     if k not in ('args', 'kwargs', 'self', '__class__', 'frame', 'frame_args')}
+        child = _get_latest_child(frame)
+
+        # we'll save hparams automatically (renamed to module_arguments)
+        for arg, val in init_args.items():
+            # don't overwrite something already set
+            if hasattr(child, arg):
+                log.warning(f'init argument `{arg}` was skipped while auto `hparams` registering,'
+                            ' because ut match already existing attribute of this class.')
+                continue
+            setattr(child, arg, val)
+
+        # set module_arguments in child
+        # skip `hparams` to uncycle with property
+        setattr(child, 'module_arguments', [k for k in init_args if k != 'hparams'])
+
+    def get_hyper_params(self, save_mode: True) -> dict:
+        """Copy the actual values from model according the list
+
+        Args:
+            weights_only: saving model weights only
+            save_mode: drop all init argument which are not primitives
+
+        Return:
+             structured dictionary
+        """
+        module_args = {}
+        for k in self.module_arguments:
+            val = getattr(self, k)
+            if save_mode and not is_picklable(val):
```

Comment:
@williamFalcon  I 'm not quite understanding why pickling all arguments is a good idea. Say you want to write a module consisting of several submodules - it can be seq2seq, autoencoder, gan etc. One way to implement such module is following

```
class GANModule(LightningModule):
    def __init__(self, hparams, generator: Module, discriminator: Module):
	self.g = generator
	self.d = discriminator
	...
```
Now if you'll follow this approach you'll checkpoint your model twice.

Other way is to create both models in a `GANModule`, by just passing hparams, so `GanModule` would define methods like `build_generator()`. But this seems inferior, as it breaks [inversion-of-control](https://en.wikipedia.org/wiki/Inversion_of_control) principle, some problems are
- Harder to write unittests
- If you want to experiment with many architectures, your GANModule can become too cluttered, because of methods that will create submodules

Reply:
in the current PR, we would not save generator and discriminator to the checkpoint. Instead, the user will have to pass those in during loading.

```python
GanModule.load_from_checkpoint(PATH, generator=MyNewGeneratorThatIDidntTrainWith, discriminator=MyOldDisc)
```